controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+// File:          teleport_controller.cpp
+// Date:          23/03/2021
+// Description:   Method to teleport robot around the playing field for the purposes of
+//                    gathering training data
+// Author:        Lachlan Court
+// Modifications:
+
+
+#include <RobotisOp2GaitManager.hpp>
+#include <RobotisOp2MotionManager.hpp>
+#include <webots/Supervisor.hpp>
+#include <iostream>
+#include <math.h>
+#include <time.h> 
+
+// All the webots classes are defined in the "webots" namespace
+using namespace webots;
```

Comment:
`using namespace std` is a horrendous thing and should be avoided at all costs. This, I believe, is mainly because many other libraries may implement functions in global namespace (horrible practice) that will clash with standard library functions.

I usually extend this to other single name `using`s (e.g. `using namespace eigen`). Mainly because it makes it more annoying to know where functions are coming from without really saving you much of anything.

If the namespace you are `using` is long and/or nested then using a `using` will make your code more readable, but we rarely, if ever, do this in a header file.

@TrentHouliston may have more thoughts on this

Reply:
Are you proposing I don't use the namespace in this way? This was just from the sample code in the webots repo and it worked so I kept it aha, I can work out which specific functions I'm using though and call webots::Motor or w/e if that would be more suitable

Reply:
Personally, I think that would be good

Reply:
Using namespace is almost always bad, there are some very very few use cases where it's a good idea. But if you are good enough to know when those are, you probably don't need c++ tips so much anymore :P

Basically, the problem with it is that it drags symbols into your current namespace that can conflict with symbols you're defining yourself. E.g. if you wrote yourself a nice function named `count`, and had `using namespace std` the compiler wouldn't know if you wanted to refer to your version of `count`, or the standard libraries version of `count`. The error it gives you, in this case, is very hard to read if you haven't read it before and can mess up a few hours of your time.

This is especially bad when you do using namespace in headers because it doesn't just pollute your code you're writing now, but potentially anyone who includes that header, even transitively. So you can add using namespace in your header happily coding away and a weird esoteric error can crop up on the other side of the codebase with a message like "Ambiguous symbol here!` and when you are causing compile errors in code you're not even writing you'll have a bad day (or potentially whoever includes your header in the future will.

Remember kids, always code as if the person maintaining your code is a violent psychopath who knows where you live. Because they will be, and it'll probably just be you a few months from now.

Hierarchy of bad using/using namespace ideas

1. 😭 `using namespace` in a header
  - never do this you monster, you're screwing over yourself and everyone else!
2. 😢 `using namespace` in a cpp file 
  - better than doing this in a header so at least you're not screwing over others  but you're still screwing over yourself and future people who work on this code
3. 😞 `using x = ...` in a header
  - don't do this unless you really want others to access the thing you're using in other files like this
5. 😐 `using namespace` in a function 
  - This is the one time that I won't yell at you for doing a using namespace it doesn't pollute the scope so much (only that function is affected) and nobody likes typing `std::chrono::duration_cast<std::chrono::duration<double>(std::chrono::steady_clock::now()).count()` when they can do `duration_cast<duration<double>>(steady_clock::now()).count();` just with a little `using namespace std::chrono;`
4. 😃 `using x = ...` in a cpp file 
  - this one is super useful and we do it all the time, go nuts

Reply:
Awesome thanks for that! I'll keep it in mind :)